main.nf

```diff
@@ -0,0 +1,527 @@
+}
+
+if(!gzi){
+  process preprocess_gzi {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaToGziCh
+
+    output:
+    file("*.gz.gzi") into gziCh
+
+    script:
+    """
+    bgzip -c -i ${fasta} > ${fasta}.gz
+    """
+  }
+}
+
+//channel to collect all necessary reference files
+fastaChannel = Channel.from(fastaCh).mix(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh).collect()
+
+/********************************************************************
+  process preprocess_bam
+  Takes care of the read group line.
+********************************************************************/
+
+process preprocess_bam{
+
+  tag "${bam}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  file(bam) from bamChannel
+
+  output:
+  set file("ready/${bam}"), file("ready/${bam}.bai") into completeChannel
+
+  script:
+  """
+  mkdir ready
+  [[ `samtools view -H ${bam} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { picard AddOrReplaceReadGroups \
+  I=${bam} \
+  O=ready/${bam} \
+  RGID=${params.rgid} \
+  RGLB=${params.rglb} \
+  RGPL=${params.rgpl} \
+  RGPU=${params.rgpu} \
+  RGSM=${params.rgsm};}
+  cd ready ;samtools index ${bam};
+  """
+}
+
+
+fastaChannel.map{fasta, bed, fai, fastagz, gzfai, gzi -> tuple (1, fasta, bed, fai, fastagz, gzfai, gzi)}
+            .set{all_fa};
+
+completeChannel.map { bam, bai -> tuple(1, bam, bai) }
+               .set{all_bam};
+
+all_fa.cross(all_bam)
+      .set{all_fa_bam};
+
+/********************************************************************
+  process make_examples
+  Getting bam files and converting them to images ( named examples )
+********************************************************************/
+
+    process make_examples{
+
+      tag "${bam[1]}"
+
+      input:
+      set file(fasta), file(bam) from all_fa_bam
+
+      output:
+      set file("${fasta[1]}"),file("${fasta[1]}.fai"),file("${fasta[1]}.gz"),file("${fasta[1]}.gz.fai"), file("${fasta[1]}.gz.gzi"),val("${bam[1]}"), file("shardedExamples") into examples
+
+      script:
+      """
+      mkdir logs
+      mkdir shardedExamples
+      dv_make_examples.py \
+      --cores ${task.cpus} \
+      --sample ${bam[1]} \
+      --ref ${fasta[1]}.gz \
+      --reads ${bam[1]} \
+      --regions ${fasta[2]} \
+      --logdir logs \
+      --examples shardedExamples
+      """
+    }
+/********************************************************************
+  process call_variants
+  Doing the variant calling based on the ML trained model.
+********************************************************************/
+
+process call_variants{
+
+  tag "${bam}"
+
+  input:
+  set file(fasta),file("${fasta}.fai"),file("${fasta}.gz"),file("${fasta}.gz.fai"), file("${fasta}.gz.gzi"),val(bam), file("shardedExamples") from examples
+
+  output:
+  set file(fasta),file("${fasta}.fai"),file("${fasta}.gz"),file("${fasta}.gz.fai"), file("${fasta}.gz.gzi"), val(bam), file('call_variants_output.tfrecord') into called_variants
+
```

Comment:
are really used all these $fasta.* files? I don't see any reference in the task.